spec/consensus/fork-detection.md

```diff
@@ -0,0 +1,45 @@
+
+## Identifying faulty validators
+
+Once an honest lite client detects a valid fork it needs to submit it to an honest full node for further processing so that faulty validators can be detected and punished. For the purpose of this document we assume that a lite client is able to submit a proof of fork to an honest full node in reasonable time (how we can ensure that will be addressed in a separate document).
+
+For the purpose of this document we assume existence of the following information for each commit: hash of the block, height, round, validator set and a set of signatures by the processes in the given validator set. An honest full node starts the procedure that detects misbehaving validators from the following information: for some height h, there are two conflicting commits C1 and C2, where C1 is a commit from the main chain. The detection procedure works like this:
+
+if there are processes from the set signers(C2) that are not part of C1.valset, they misbehaved as they are signing protocol messages in heights they are not validators. In this case commit C2 is a self-contained evidence of misbehavior for those processes that can be simply verified by every honest validators without additional information.
+For processes from the set signers(C2) that are part of C1.valset we need additional checks:
+
+if C1.round == C2.round, and some processes signed different precommit messages in both commits, then it is an equivocation misbehavior. Similarly as above, commits C1 and C2 are an evidence of misbehavior. Note that in this case there might be additional misbehavior, but its detection require more complex detection procedure that we explain next.
+
+if C1.round != C2.round we need to run full detection procedure. We assume (in the first version) that the full detection procedure is executed by a centralized, trusted component called monitor that will be processing proof of forks and identifying faulty processes (and optionally generating evidences of misbehavior). Monitor is an honest full node that runs on the main chain. Monitor is triggered by receiving two conflicting commits where C1.round != C2.round. The procedure starts by declaring processes that are in the set C1.valset as suspected. This can be for example done by posting a special transaction on the main chain. After suspected processes are declared, they are obliged to submit vote sets (prevote and precommit messages) for the given height for the subset of rounds [C1.round, C2.round] to the monitor process within MONITOR_RESPONSE_PERIOD. Validators that do not submit its vote sets within this period are considered faulty (Q: how can we create verifiable proof of absence of vote sets?).
```

Comment:
> centralized, trusted component

this destroyed the idea of distributed/decentralized ledger. I'd say everybody can run a monitor node

Reply:
> Q: how can we create verifiable proof of absence of vote sets?

what if instead of sending it to concrete monitor process, we require publishing vote sets on chain? that way they become easily verifiable and publicly accessible.

Reply:
re centralised monitor, we agreed in the meantime that we will be running monitor as on-chain module. So it will be run by everyone but we will have consistent output what are faulty validators. And yes, vote sets will be published on chain as part of this procedure. It's my fault for not having written this yet; will try to do it in the following days.

Reply:
> It's my fault for not having written this yet; 

Don't worry about it too much. I've only started working on it.

> will try to do it in the following days.

thanks!